File: server_old/metadata/Spotify.ts
import { ulid } from "ulid";
import { Media, Spotify } from "../server";
import { Album, Artist, ExtendedArtist, MetadataProviders, Track } from "../struct";

export default async function getSpotifyArtist(id: string): Promise<ExtendedArtist> {
  const artist = (await Spotify.call(async () => await Spotify.api.getArtist(id)))?.body;
  if (!artist) return;
  const newArtist: ExtendedArtist = {
    ...constructArtistFromSpotify(artist),
    albums: [],
  };
  let albumOffset = 0;
  await (async function newAlbumSet() {
    const fullAlbums = await Spotify.call(
      async () =>
        (
          await Spotify.api.getArtistAlbums(artist.id, {
            limit: 50,
            offset: albumOffset,
            include_groups: "album,single,appears_on",
          })
        ).body.items
    );
    fullAlbums
      .filter((a) => a.album_type !== "compilation")
      .forEach((a) => {
        if (!newArtist.albums.find((l) => l.name == a.name))
          newArtist.albums.push(constructAlbumFromSpotify(a));
      });
    if (fullAlbums.length == 50) {
      albumOffset += 50;
      await newAlbumSet();
    }
  })();

  await Promise.all(
    newArtist.albums.map(async (a) => {
      let trackOffset = 0;
      const songs: SpotifyApi.TrackObjectSimplified[] = [];
      await (async function newTrackSet(time: number) {
        await new Promise((r) => setTimeout(r, time));
        try {
          const albumSongs = await Spotify.call(
            async () =>
              (
                await Spotify.api.getAlbumTracks(a.uuid, {
                  limit: 50,
                  offset: trackOffset,
                })
              ).body.items
          );
          songs.push(...albumSongs.filter((s) => s.artists.find((sa) => sa.id == artist.id)));
          if (albumSongs.length == 50) {
            trackOffset += 50;
            await newTrackSet(500);
          } else {
            a.tracks = songs.map(constructTrackFromSpotify);
          }
        } catch {
          await newTrackSet(1000);
        }
      })(1);
    })
  );

  // attempts to sort out singles that are separate from their albums
  /*newArtist.albums
    .filter(
      (a) =>
        a.type == "single" &&
        a.tracks.length == 1 &&
        a.tracks[0].title == a.name
    )
    .forEach((a) => {
      if (
        newArtist.albums.find(
          (l) =>
            l !== a &&
            l.tracks.find(
              (t) => t.title == a.name && t.duration == a.tracks[0].duration
            )
        )
      )
        newArtist.albums.splice(newArtist.albums.indexOf(a), 1);
    });*/

  newArtist.albums = newArtist.albums.filter((a) => a.tracks.length);
  return newArtist;
}

export async function searchSpotifyArtists(query: string): Promise<Artist[]> {
  return (
    await Spotify.call(
      async () =>
        await Spotify.api.searchArtists(query, {
          limit: 15,
        })
    )
  ).body.artists.items.map((a) => {
    const artist = constructArtistFromSpotify(a);
    return { ...artist, status: Media.hasArtist(artist) ? 2 : 0 };
  });
}

function findImage(images: SpotifyApi.ImageObject[]) {
  return (images.find((i) => i.width == i.height) || images[0])?.url || "";
}

export function constructArtistFromSpotify(artist: SpotifyApi.ArtistObjectFull): Artist {
  return {
    id: ulid(),
    name: artist.name || "",
    url: artist.external_urls.spotify || "",
    genres: artist.genres || [],
    followers: artist.followers.total || 0,
    icon: findImage(artist.images),
    providers: {
      [MetadataProviders.Spotify]: artist.id,
    },
  };
}

export function constructAlbumFromSpotify(album: SpotifyApi.AlbumObjectSimplified): Album {
  return {
    type: (album.album_type as any) || "",
    url: album.external_urls.spotify || "",
    id: ulid(),
    name: album.name || "",
    year: Number(album.release_date.split("-")[0]) || 0,
    image: findImage(album.images),
    tracks: [],
    uuid: album.id || "",
    provider: MetadataProviders.Spotify,
  };
}

export function constructTrackFromSpotify(track: SpotifyApi.TrackObjectSimplified): Track {
  return {
    id: track.id,
    uuid: track.id,
    title: track.name || "",
    url: track.external_urls.spotify || "",
    number: track.track_number || 0,
    duration: track.duration_ms || 0,
    explicit: track.explicit || false,
  };
}

File: server/metadata/spotify.go
package metadata

import (
	"context"
	"log"

	"github.com/itzTheMeow/YTMusic/types"
	"github.com/itzTheMeow/YTMusic/util"
	"github.com/oklog/ulid/v2"
	"github.com/zmb3/spotify/v2"
	spotifyauth "github.com/zmb3/spotify/v2/auth"
	"golang.org/x/oauth2/clientcredentials"
)

var ctx context.Context
var SpotifyClient *spotify.Client

func InitSpotify() {
	ctx = context.Background()
	credentials := &clientcredentials.Config{
		ClientID:     util.UserConfig.SpotifyClientID,
		ClientSecret: util.UserConfig.SpotifyClientSecret,
		TokenURL:     spotifyauth.TokenURL,
	}
	token, err := credentials.Token(ctx)
	if err != nil {
		log.Printf("Couldn't get token for spotify: %v", err)
	} else {
		log.Print("Obtained spotify access token.")
		SpotifyClient = spotify.New(spotifyauth.New().Client(ctx, token))
	}
}

func FetchSpotifyArtist(id string) (*types.Artist, error) {
	spotifyArtist, err := SpotifyClient.GetArtist(ctx, spotify.ID(id))
	if err != nil {
		return nil, err
	}

	artist := ConstructArtistFromSpotify(*spotifyArtist)
	artist.Albums = make([]types.Album, 0)

	return &artist, nil
}

func SearchSpotifyArtists(query string) []types.Artist {
	artists := make([]types.Artist, 0)
	if SpotifyClient != nil {
		res, err := SpotifyClient.Search(ctx, query, spotify.SearchTypeArtist, spotify.Limit(15))
		if err == nil {
			for _, artist := range res.Artists.Artists {
				artists = append(artists, ConstructArtistFromSpotify(artist))
			}
		} else {
			log.Println(err)
		}
	}
	return artists
}

func findImage(images []spotify.Image) string {
	for _, img := range images {
		if img.Width == img.Height {
			return img.URL
		}
	}
	if len(images) > 0 {
		return images[0].URL
	} else {
		return ""
	}
}

func ConstructArtistFromSpotify(artist spotify.FullArtist) types.Artist {
	providers := make(map[types.MetadataProvider]string)
	providers[types.MetaProviderSpotify] = artist.ID.String()
	return types.Artist{
		ID:        ulid.Make().String(),
		Name:      artist.Name,
		URL:       artist.ExternalURLs["spotify"],
		Genres:    artist.Genres,
		Followers: int(artist.Followers.Count),
		Icon:      findImage(artist.Images),
		Providers: providers,
	}
}
func ConstructAlbumFromSpotify(album spotify.SimpleAlbum) types.Album {
	return types.Album{
		Type:     types.AlbumType(album.AlbumType),
		URL:      album.ExternalURLs["spotify"],
		ID:       ulid.Make().String(),
		Name:     album.Name,
		Year:     album.ReleaseDateTime().Year(),
		Image:    findImage(album.Images),
		Tracks:   make([]types.Track, 0),
		UUID:     album.ID.String(),
		Provider: types.MetaProviderSpotify,
	}
}
func ConstructTrackFromSpotify(track spotify.FullTrack) types.Track {
	return types.Track{
		ID:       ulid.Make().String(),
		UUID:     track.ID.String(),
		Title:    track.Name,
		URL:      track.ExternalURLs["spotify"],
		Number:   track.TrackNumber,
		Duration: int(track.TimeDuration().Milliseconds()),
		Explicit: track.Explicit,
	}
}
